addons/stock/models/stock_production_lot.py

```diff
@@ -27,6 +30,47 @@ class ProductionLot(models.Model):
     display_complete = fields.Boolean(compute='_compute_display_complete')
     company_id = fields.Many2one('res.company', 'Company', required=True, store=True, index=True)
 
+    @api.model
+    def generate_lot_names(self, first_lot, count):
+        """ This method will generate `lot_names` from a string.
+        """
+
+        # We look if the first lot contains at least one digit.
+        caught_initial_number = regex_findall(r"\d+", first_lot)
+        if not caught_initial_number:
+            raise UserError(_('The lot name must contain at least one digit.'))
+        # We base the serie on the last number found in the base lot.
+        initial_number = caught_initial_number[-1]
+        padding = len(initial_number)
+        # We split the lot name to get the prefix and suffix.
+        splitted = regex_split(initial_number, first_lot)
+        # initial_number could appear several times, e.g. BAV023B00001S00001
+        prefix = initial_number.join(splitted[:-1])
+        suffix = splitted[-1]
+        initial_number = int(initial_number)
+
+        lot_names = []
+        for i in range(0, count):
+            lot_names.append('%s%s%s' % (
+                prefix,
+                str(initial_number + i).zfill(padding),
+                suffix
+            ))
+        return lot_names
+
+    @api.model
+    def get_next_serial(self, company, product):
+        """ This method will return the next serial number to be attributed to the product
+        based on the highest generated.
+        """
+        if product.tracking == "serial":
+            last_serial = self.env['stock.production.lot'].search(
+                [('company_id', '=', company.id), ('product_id', '=', product.id)],
+                limit=1, order='name DESC')
```

Comment:
This definitely needs to be changed. If you use order='name DESC' you will have issues with repeat SNs by default once you have cases such as:
SNs used = 'a9', 'a10' => 'a9' = last serial and 'a10' will always the default "First SN" in your wizard.

Reply:
Considering generate_lot_names and user documentation showing 'SN-001' as first sn in sequence to be generated,
ordering by name seemed the best solution for real-life sn (fixed length with zero padding for the sn part).
-> checked with mgm and confirmed this way 
Set the very first sn to 'a0001' rather than 'a1' to solve this issue.
However, I adapted the code to handle full sequences or sequence like 1,2,...,9,10,...

Reply:
Good point about the real-life sn. I agree it's better have safer though because I'm sure there will be users who will vary the length of their SNs and run into this problem. I'm not sure if I understand your solution though, wouldn't it be easier to assume users will always create the SNs in order so we can always search by create_date to determine the next SN's name? Especially since it looks like we will do 2 searches whenever the last_serial initial_number[0] != '0', which will happen 9 out of 10 times?